### PR TITLE
Make public and external methods virtual to facilitate creation of custom SuperToken logic

### PR DESCRIPTION
**Pull request by d10r**
Customized versions of the SuperToken logic currently cannot just inherit `SuperToken.sol`, because it doesn't declare its methods as `virtual`.

Examples:
* In https://github.com/superfluid-finance/custom-supertokens/pull/26, `updateCode` changes.
* In https://github.com/GoodDollar/GoodProtocol/blob/master/contracts/token/superfluid/SuperToken.sol, the implementation of several methods (e.g. `transfer`) is changed.

In this PR, all public and external methods of SuperToken (and SuperfluidToken which it inherits from) are declared as `virtual` in order to facilitate creation of such customizations.

Additionally a few internal methods (for mint, burn, transfer, upgrade and downgrade) were also made virtual in order to allow changing the behaviour of that operations with less additional code.

### CODE REVIEW

File: packages/ethereum-contracts/contracts/superfluid/SuperToken.sol [hunk 1]
```diff
@@ -192,7 +193,7 @@ contract SuperToken is
      * For more context, see https://github.com/superfluid-finance/protocol-monorepo/wiki/About-ERC-777
      */
     function _transferFrom(address spender, address holder, address recipient, uint amount)
-        internal returns (bool)
+        internal virtual returns (bool)
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/superfluid/SuperToken.sol]: We are doing a much bigger surgery than I heard of.

This is code smell and anti-pattern, in general one should not change the behaviour of these "_*" functions for it is crafted logic for the reusing site.

Instead, a more idiomatic OO way (not really promoting it anymore, but within the conext of this codebase) is to override the site that uses those code, rewiring the logic etc.

Comment (hellwolf) [reply packages/ethereum-contracts/contracts/superfluid/SuperToken.sol]: let's discuss it separately tomorrow.

**Review by hellwolf — CHANGES_REQUESTED** [review]
Code organization principles:

- `_foo` internal function as building blocks for reusability.
  - they should not be altered by subclasses.
  - "private" scope is generally avoided for simplicity, since test-harness may benefit from accessing them directly by subclassing it. This is often the case for our test cases.
  - The building blocks become "monadic", if it calls other abstract virtual functions. The naming convention for these functions are "_doSomeDamage".
- `external` functions can be created through assembling these internal logic together.
  - `public` functions is a shortcut used in code base to combine the definition of  `_foo` and `foo`. This style should be gradually rid of from this code base, in favor of separation of definitions from day 0.


There are generally two ways of code logic reassembling:

```
function foo() external virtual override {
   // inject some logic before...
   super.foo();
   // inject some logic after...
}
```

or a complete rewrite of `foo` using `_foo` etc.

File: packages/ethereum-contracts/contracts/superfluid/SuperToken.sol (hunk 1 above)
Comment (d10r) [reply packages/ethereum-contracts/contracts/superfluid/SuperToken.sol]: changed to only have external and public methods virtual

**Review by hellwolf — APPROVED** [review]

**Review by hellwolf — APPROVED** [review]
I still find it a bit stinky to be able to override so many things, but it is at least better than being able to override internal logic blocks.

**Pull request merged**